Fix Ollama GPT-OSS streaming with 'thinking' field

colesmcintosh opened the pull request:
## Title

Fix Ollama GPT-OSS streaming with 'thinking' field

## Relevant issues

Fixes #13340

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🐛 Bug Fix

## Changes

### Problem
Ollama GPT-OSS models were failing with `APIConnectionError: Unable to parse ollama chunk` when streaming responses contained a `'thinking'` field with empty `'response'`. The chunk parser didn't handle this specific case, causing streaming to fail.

### Solution
- Added handling for chunks containing `'thinking'` field with empty `'response'`
- These chunks are treated as intermediate chunks that don't contain user-facing content
- Allows streaming to continue until actual response content arrives

### Code Changes
- Modified `OllamaTextCompletionResponseIterator.chunk_parser()` in `litellm/llms/ollama/completion/transformation.py`
- Added condition to handle `"thinking" in chunk and not chunk["response"]`
- Returns empty `GenericStreamingChunk` for these intermediate chunks

### Tests Added
- `test_chunk_parser_with_thinking_field()`: Tests the exact problematic chunk from the issue
- `test_chunk_parser_normal_response()`: Ensures normal chunks still work
- `test_chunk_parser_done_chunk()`: Verifies done chunks work correctly

### Verification
Tested with the exact chunk from the error:
```json
{'model': 'gpt-oss:20b', 'created_at': '2025-08-06T14:34:31.5276077Z', 'response': '', 'thinking': 'User', 'done': False}
```

Result: Successfully parsed without errors, returns empty text chunk allowing stream to continue.

colesmcintosh committed:
Fix Ollama GPT-OSS streaming with 'thinking' field

- Handle chunks containing 'thinking' field with empty 'response'
- Treat these as intermediate chunks that don't contain user content
- Add comprehensive tests for chunk parsing scenarios
- Resolves APIConnectionError for GPT-OSS model streaming

Fixes #13340

colesmcintosh committed:
Fix unclosed aiohttp client session warnings during concurrent requests

Fixed MyPy type error in Ollama completion transformation where finish_reason
was set to None instead of expected string type. Changed finish_reason=None to
finish_reason="" to match GenericStreamingChunk TypedDict requirements.

Also updated corresponding test to expect empty string instead of None.

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/ollama-gpt-oss-thinking-field

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/ollama-gpt-oss-thinking-field

krrishdholakia commented on litellm/llms/ollama/completion/transformation.py:
```diff
@@ -459,6 +459,15 @@ def chunk_parser(self, chunk: dict) -> GenericStreamingChunk:
                     finish_reason="stop",
                     usage=None,
                 )
+            elif "thinking" in chunk and not chunk["response"]:
```
@colesmcintosh can we return a ModelResponseStream instead? 

This way we can include the reasoning content in the response, and allow it to be displayed on chat ui's like OpenWebUI - See openrouter implementation - https://github.com/BerriAI/litellm/blob/184687157e053e106721f14b25cb22f896efe499/litellm/llms/openrouter/chat/transformation.py#L108

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/ollama-gpt-oss-thinking-field

colesmcintosh committed:
Enhance chunk parsing for Ollama streaming responses

Updated the chunk_parser method to return a ModelResponseStream when handling 'thinking' field content, allowing UIs to render reasoning information. Adjusted tests to verify the new behavior, ensuring that reasoning content is correctly returned in the response.

colesmcintosh committed:
merge from upstream

colesmcintosh committed:
Revert "merge from upstream"

This reverts commit e6ca91869a7eef75b0e31e38788b0a0ccf38e377.

benlangfeld commented:
Hey everyone, is there an ETA on shipping this? It's not possible to use the gpt-oss model behind LiteLLM without this fix.

krrishdholakia commented:
@benlangfeld Can you confirm the fix works for you?

zigelboim-misha commented:
@krrishdholakia It should work in the `LiteLLM Proxy` too? Or not yet?

colesmcintosh commented:
> @krrishdholakia It should work in the `LiteLLM Proxy` too? Or not yet?

It should work throughout - let me know if you're running into an error
